lib-new/GraphDatabase.coffee

```diff
@@ -0,0 +1,316 @@
+$ = require 'underscore'
+assert = require 'assert'
+{Error} = require './errors'
+lib = require '../package.json'
+Node = require './Node'
+Relationship = require './Relationship'
+Request = require 'request'
+Transaction = require './Transaction'
+URL = require 'url'
+
+
+module.exports = class GraphDatabase
+
+    # Default HTTP headers:
+    headers:
+        'User-Agent': "node-neo4j/#{lib.version}"
+        'X-Stream': 'true'
+
+    constructor: (opts={}) ->
+        if typeof opts is 'string'
+            opts = {url: opts}
+
+        {@url, @headers, @proxy, @agent} = opts
+
+        if not @url
+            throw new TypeError 'URL to Neo4j required'
+
+        # TODO: Do we want to special-case User-Agent? Blacklist X-Stream?
+        @headers or= {}
+        $(@headers).defaults @constructor::headers
+
+    http: (opts={}, cb) ->
+        if typeof opts is 'string'
+            opts = {path: opts}
+
+        {method, path, headers, body, raw} = opts
+
+        if not path
+            throw new TypeError 'Path required'
+
+        method or= 'GET'
+        headers or= {}
+
+        # TODO: Would be good to test custom proxy and agent, but difficult.
+        # Same with Neo4j returning gzipped responses (e.g. through an LB).
+        req = Request
+            method: method
+            url: URL.resolve @url, path
+            proxy: @proxy
+            headers: $(headers).defaults @headers
+            agent: @agent
+            json: body ? true
+            gzip: true  # This is only for responses: decode if gzipped.
+
+        # Important: only pass a callback to Request if a callback was passed
+        # to us. This prevents Request from doing unnecessary JSON parse work
+        # if the caller prefers to stream the response instead of buffer it.
+        , cb and (err, resp) =>
+            if err
+                # TODO: Do we want to wrap or modify native errors?
+                return cb err
+
+            if raw
+                # TODO: Do we want to return our own Response object?
+                return cb null, resp
+
+            if err = Error._fromResponse resp
+                return cb err
+
+            cb null, _transform resp.body
+
+        # Instead of leaking our (third-party) Request instance, make sure to
+        # explicitly return only its internal native ClientRequest instance.
+        # https://github.com/request/request/blob/v2.53.1/request.js#L904
+        # This is only populated when the request is `start`ed, so `start` it!
+        req.start()
+        req.req
+
+    cypher: (opts={}, cb, _tx) ->
+        if typeof opts is 'string'
+            opts = {query: opts}
+
+        if opts instanceof Array
+            opts = {queries: opts}
+
+        {queries, query, params, headers, lean, commit, rollback} = opts
+
+        if not _tx and rollback
+            throw new Error 'Illegal state: rolling back without a transaction!'
```

Comment:
I think we also need to check `if rollback and _tx and not _tx._id` and early exit if it's true. Otherwise we'll throw an error. This should be valid code even if it amounts to a no-op:
```
tx = DB.beginTransaction()
tx.rollback()
```